File: solutions/python/pig-latin/2/pig_latin.py
VOWELS = 'aeiou'


def translate(text):
    return ' '.join(map(translate_word, text.split()))


def translate_word(word):
    if word[0] in VOWELS or word[:2] in ['xr', 'yt']:
        result = word
    else:
        head, tail = separate_leading_consonants(word)
        result = tail + head
    return result + 'ay'


def separate_leading_consonants(word):
    if word.startswith('y'):
        tail_index = 1
    else:
        tail_index = find_first_vowel(word)
        if word[tail_index-1 : tail_index+1] == 'qu':
            tail_index += 1
    return word[:tail_index], word[tail_index:]


def find_first_vowel(word):
    for i, c in enumerate(word):
        if c in VOWELS + 'y':
            return i

File: solutions/fsharp/pig-latin/1/PigLatin.fs
module PigLatin

open System.Text.RegularExpressions

let splitPrefix (word: string) : (string * string) option = 
    [
        @"(^xr.*|^yt.*)()";
        @"^([^aeiou]*qu)(.*)";
        @"^([^aeiou]+)([aeiouy].*)"
        @"^([^aeiou]+)(.*)";
    ] 
        |> List.map (fun pattern -> Regex.Match(word, pattern))
        |> Seq.tryFind (fun _match -> _match.Success)
        |> Option.map (fun _match -> (_match.Groups[1].Value, _match.Groups[2].Value))

let translateWord word = 
    match splitPrefix word with
    | Some(head, tail) -> tail + head + "ay"
    | None -> word + "ay"

let translate (input: string) : string =
    input.Split(' ') |> Array.map translateWord |> String.concat " "
